lib/identity_cache/cache_fetcher.rb

```diff
@@ -1,7 +1,58 @@
+# frozen_string_literal: true
+
+require 'securerandom'
+
 module IdentityCache
   class CacheFetcher
     attr_accessor :cache_backend
 
+    EMPTY_HASH = {}.freeze
+
+    class FillLock
+      FILL_LOCKED = :fill_locked
+      FAILED_CLIENT_ID = 'fill_failed'
+
+      class << self
+        private :new
+
+        def build(client_id:, data_version:)
+          new([FILL_LOCKED, client_id, data_version])
+        end
+
+        def from_cache(cache_value)
+          raise ArgumentError unless cache_value?(cache_value)
+          new(cache_value)
+        end
+
+        def cache_value?(cache_value)
+          cache_value.is_a?(Array) && cache_value.length == 3 && cache_value.first == FILL_LOCKED
+        end
+      end
+
+      attr_reader :client_id, :data_version, :cache_value
+
+      def initialize(cache_value)
+        @cache_value = cache_value
+        _, @client_id, @data_version = cache_value
```

Comment:
Is there a significant performance benefit to initializing this with an array and storing both the array and its components? It seems like it'd be cleaner to:
```ruby
attr_reader :client_id, :data_version

def initialize(client_id:, data_version:)
  @client_id = client_id
  @data_version = data_version
end

def cache_value
  [FILL_LOCKED, @client_id, @data_version]
end
```

Reply:
We already need to support initializing it from an array when it comes from the cache and storing the array simplified equality checking.

I don't think there would be a significant performance impact from not storing the array, so I'll change it to what you suggested.